### PR TITLE
Handle SurfaceView in a VirtualDisplay

### PR DESCRIPTION
**Pull request by blasten**
Fixes https://github.com/flutter/flutter/issues/103630
Fixes https://github.com/flutter/flutter/issues/105882

SurfaceViews punch a whole in the window to allow the surface to be displayed. 
https://github.com/flutter/engine/pull/31468 caused an issue where platform views that embed a SurfaceView cannot be composed correctly in Flutter. This used to work previously thanks to VirtualDisplay although it wasn't a documented behavior. 

This PR brings back the VirtualDisplay for views that can do their own drawing without the Android view.  For example, a plugin like flutter-mapbox-gl uses SurfaceView to issue GL commands to this surface directly.

To review this PR, you would need to compare the differences between this patch and https://github.com/flutter/engine/pull/31468/files.

### CODE REVIEW

**Comment by penhorwood** [comment]
Is this Linux checked being worked on?  Seems like a lot of people are waiting on this fix.

**Comment by penhorwood** [comment]
I have followed most of the related links to this pull request.  Is there anything we can do to help get this completed?

**Comment by blasten** [comment]
@penhorwood I was waiting for https://github.com/flutter/engine/pull/33574 to land. This fix needs an integration test, so this issue doesn't reoccur in the future.

**Comment by sirmamedical** [comment]
Any progress on releasing this fix? 
We are stuck with flutter 2.10.5 and with the compatible plugin versions only. Most of the plugin updates are not possible.

**Comment by penhorwood** [comment]
Seems we all lost another week to this issue.  Are we still waiting on a integration test?

**Comment by penhorwood** [comment]
Getting closer.  I like the changes you made to the patch.  What happens next?

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 14.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

**Comment by skia-gold** [comment]
Gold has detected about 21 new digest(s) on patchset 18.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

**Comment by blasten** [comment]
I'm testing TalkBack in the Google maps example app with Flutter 2.8.1 and the current master + this patch. It looks like moving Google maps to the regular view hierarchy (as it's in Flutter master and stable) fixed TalkBack, and applying this patch regresses that. 

I'm still debugging to see what else can be done since there was a P1 recently about Talkback not working in Google Maps, I expect this to be an issue that needs to be addressed before landing this patch.

cc @stuartmorgan

**Comment by skia-gold** [comment]
Gold has detected about 21 new digest(s) on patchset 19.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

**Comment by blasten** [comment]
@stuartmorgan FYI: https://github.com/flutter/engine/pull/33599/commits/8f821f6f609e0d7c49a7ed3d0c12ee27d214edf9

**Comment by skia-gold** [comment]
Gold has detected about 20 new digest(s) on patchset 20.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: These limitations seem pretty severe.

What are we solving with this? Just unbreaking mapbox? Are there lots of use cases for SurfaceView inside a platform view? Could we be doing something better here and just giving mapbox a GL context to work with?

What views need to call invalidate? Just TextureViews?

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: Of course, if we did give plugins some way to directly talk to GL, we'd have to have them do some heavier lifting around a11y and input anyway...

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:981]: This seems like we should just clamp to the display size and print a warning that we did that...

**Review by dnfield — COMMENTED** [review]
I haven't carefully read the whole patch yet, some initial questions.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: But rather than having SurfaceView or TextureView in a platform view, could we just give plugins a GL context to draw into?

Alternatively, could the plugin just use the `TextureRegistry` API instead? That should generally perform better than platform views - but again, I'm not sure I really understand all the requirements. For a WebRTC plugin, I'd imagine they want to treat that more like how exoplayer works rather than like a platform view though...

Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: I would actually love to see is an official API that allows to project a View into a GL Texture while the view is in the view hierarchy. Some challenges:
1. it must support embedded TextureViews like the Google Maps View.
2. it must support embedded SurfaceViews like mapbox or WebRTC views. 
3. WebView is a special view that does its own thing

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: My suggestion here is not that Android somehow give us an API to do this, but that we ask the mapbox and webRTC plugin teams to refactor their plugins to do this.

Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: > But rather than having SurfaceView or TextureView in a platform view, could we just give plugins a GL context to draw into?

Yes, they can today. :) I'd be great if they can go down that route. 

Part of the issue is that plugins maintainers don't own the core component, so they are constrained by the public API of these components, but if they could build or adapt it for Flutter, they could maybe just build it using the Flutter framework directly, or Flutter's texture driven by an external raster thread, but use Flutter a11y API, and Flutter widgets. It's a spectrum of options, but it all boils down to how much they can afford.

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: Adding this API back gives us a relatively large API surface to maintain that has some severe known issues. I'd like to think we're driving people away from this API and towards APIs that will actually perform well.

How long ago did we remove this API? How many versions of Flutter exist without it? I think that might be the last piece I'm missing for context here. If this was only recently dropped and we're trying to quickly add it back it makes more sense to me, but it'd probably still be good to add it back as deprecated or something.

Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: > If this was only recently dropped and we're trying to quickly add it back it makes more sense to m

Yes, that is exactly the case. It was removed in https://github.com/flutter/engine/pull/31468/files, and it's currently in Flutter 3.0.x.

Comment (josephcrowell) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: There are many projects requiring the VirtualDisplay functionality. It involves every map view plugin including Google's, video player surfaces and some plugins with camera views. 

Many are linked from: https://github.com/flutter/flutter/issues/103630

I would suggest pushing through this revert until a working alternative with documentation on use is created. Plugin maintainers would gladly switch to a new working API if it existed, but there's no alternative to push them on to. This API was removed in the 3.0.0 major release and everything has remained broken since them for all plugins that utilize Android SDKs with native GL surfaces.

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:65]: I'm not opposed to relanding this. I wanted to make sure I understand the context.

That said, the video_player package does _not_ use this functionality - it uses external textures, which are much more efficient. We should try to find ways to drive plugins in more performant directions - I think it's pretty likely that platform views are never going to be the right solutino for plugins that want to draw to native GL surfaces.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:91]: Is this TODO still relevant?

If so we should have a bug filed explaining the context. Given the time and nature of this reland I suspect it's not relevant.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:579]: here and elsewhere: These logs seem to violate the style guide - they seem to be debugging information that might be useful to an engine developer but don't really mean a whole lot to a developer. They should be made actionable to end developers (did a plugin forget to make an important API call here?) or removed.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:981]: (I'm mainly concerned that the current failure mode is a potential crash on user devices that never shows up for the developer, as opposed to a potential resizing that happens on user devices)

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:278]: This should have a bug filed and linked.

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:313]: ditto (probably same bug)

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:361]: ouch.

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:382]: This is an old issue and the Andorid SDK has long since been updated a few times over. If we can get rid of some of the reflection here we should.

**Review by dnfield — COMMENTED** [review]
Mostly looks good - a few nits and a few I-hope-we-can-do-these.

I still think that ideally we'd be steering people who want to embed a Texture or SurfaceView to instead use external texture composition (like video_player does), since that is _probably_ what they want and will definitely perform better (and ideally make it a bit clearer about a11y and input responsibilities instead of a bunch of weird hacks that have really rough edges). I'm saying this more to solicit ideas about documentation ideas for this right now than to say we shouldn't unbreak plugins that got broken by dropping VirtualDisplay support.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: done

**Comment by skia-gold** [comment]
Gold has detected about 25 new digest(s) on patchset 22.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (josephcrowell) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: How would you get an activity provided by a native Android SDK that includes a widget which is a GL surface over to the Android side? Plugin developers in these cases aren't drawing a GL surface themselves, they're utilizing an SDK that does it for them. It would be wonderful to be able to use these SDKs without having stutter and jank.

Could you point me towards some documentation on piping GL surfaces straight over? In the Mapbox case, the widget in question that is usually inserted in to an activity is called a MapView https://docs.mapbox.com/android/maps/examples/simple-map-view/ and I'm pretty sure that all other map SDKs provide something similar without providing direct access to the underlying GL surface.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: good catch. I found a bug in the dispose method. I fixed it in https://github.com/flutter/engine/pull/33599/commits/f0d1029848a99378e59b6e3317b98f1bc53567eb

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:91]: I took this from the previous code, could we do the clean up in a separate PR?

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:278]: could we do this in a separate PR?

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:313]: ditto

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:361]: yeah.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:91]: I'd be fine with just deleting the TODO here, but don't consider this blocking. Filed https://github.com/flutter/flutter/issues/106443 to track.

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:278]: https://github.com/flutter/flutter/issues/106443

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:382]: similar comment. This could be a separate PR

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:582]: Filed https://github.com/flutter/flutter/issues/106444

**Comment by skia-gold** [comment]
Gold has detected about 25 new digest(s) on patchset 23.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: For MapBox, instead of using `MapView` I believe you'd use `MapSurface` (https://docs.mapbox.com/android/maps/api/10.6.0/mapbox-maps-android/com.mapbox.maps/-map-surface/). 

You may be limited by particular SDKs, but you can also file feature requests in those SDKs to provide access to things like this - if they support SurfaceView/TextureView they should generally support rendering to a Surface itself though.

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: You can see how VideoPlayer does it here: https://github.com/flutter/plugins/blob/main/packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java which uses the texture registry instead of a platform view.

File: shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java:382]: https://github.com/flutter/flutter/issues/106449

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: Filed https://github.com/flutter-mapbox-gl/maps/issues/1098 for MapBox specifically.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: I did this in https://github.com/flutter/engine/commit/5a2c81b3d4613c4410c4237d331aeba865cdbd56

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: Yes. Ideally, the MapBox plugin uses Surface directly, and the Texture widget. There's a nice article about this: https://medium.com/@german_saprykin/opengl-with-texture-widget-f919743d25d9 

/cc @felix-mittermeier

Comment (josephcrowell) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: MapView uses SurfaceView and TextureView https://github.com/mapbox/mapbox-maps-android/blob/main/sdk/src/main/java/com/mapbox/maps/MapView.kt

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: > MapView uses SurfaceView and TextureView https://github.com/mapbox/mapbox-maps-android/blob/main/sdk/src/main/java/com/mapbox/maps/MapView.kt

Right, it's not best option for flutter Olivia

**Review by dnfield — APPROVED** [review]
Lgtm

**Comment by skia-gold** [comment]
Gold has detected about 22 new digest(s) on patchset 25.
View them at https://flutter-engine-gold.skia.org/cl/github/33599

**Comment by j1mmyto9** [comment]
When will this PR be released? Why 3.0.3 doesn't wait for it? We have been waiting for so long

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java
Comment (josephcrowell) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: Would it be possible to pipe the entire activity including overlay widgets through using the texture registry?

Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: No.  Just the surface

**Pull request merged**
